[Discussion] Add yarp::os::getEnvironment(), yarp::os::setEnvironment() and yarp::os::unsetEnvironment() functions and deprecate NetworkBase version.

drdanz opened the pull request:
This PR is a "Discussion PR" (which is a term I just made up), I will not merge this without a proper discussion.

-----------

### Libraries

#### `os`

* Add `yarp::os::getEnvironment()`, `yarp::os::setEnvironment()` and `yarp::os::unsetEnvironment()` functions.

##### `Network`

* `getEnvironment()` is now deprecated in favour of `yarp::os::getEnvironment()`
* `setEnvironment()` is now deprecated in favour of `yarp::os::setEnvironment()`
* `unsetEnvironment()` is now deprecated in favour of `yarp::os::unsetEnvironment()`

-----------

As discussed in several occasions (#1853), there is no reason for having these static methods in `yarp::os::Network`, it is actually misleading and a source of circular dependencies. `Network` should be a class related to "YARP Network", and the system environment has nothing to do with that.

This patch moves these methods in the `yarp::os` namespace and deprecates the `yarp::os::Network` and `yarp::os::NetworkBase` versions

It's quite late for the release, but this should be a rather "safe" change, but there is a deprecation involved.

In #1853 @Nicogene proposed to move these methods in the `yarp::conf::environment` namespace, but these methods depend on functions that are currently in `yarp::os` and that depend on stuff in `yarp::os::impl` therefore it is not so easy to move them outside `yarp::os`.
Therefore, even if this is probably not an ideal solution, I suggest to move these methods in the `yarp::os` namespace, removing in this way the circular dependencies. AFAIK, here is no c++ proposal for handling the environment variables yet, therefore the `yarp::conf` version could be reserved in case there will be one in the future, and could have the same API.

CC @pattacini, @vtikha, @traversaro, @randaz81, @Nicogene, @elandini84, @lornat75, @robotology/yarp-developers 

Fixes #1853

drdanz commented:
If you don't want to deprecate the `Network` methods right now, we could just add these functions and use them internally, without deprecating the old ones, and maybe deprecate them in the next release cycle

drdanz commented:
We should also investigate whether this is still necessary:

https://github.com/robotology/yarp/blob/f46557b46780f23502f59f902ef78fe1637a94af/src/libYARP_os/src/yarp/os/impl/PlatformStdlib.h#L28-L52

pattacini commented:
Ok for me 👍🏻

We could also proceed with the deprecation of old methods already at this early stage as I guess they're not so ubiquitous in the user codebase.

I don't have particular concerns about that.

drdanz commented:
I created a separate PR where I moved these methods in the `yarp::conf::environment` namespace, and I removed also `yarp::os::getenv` (which doesn't have any sense, since there is a `std::getenv` method in c++98.

The result is that the code is cleaner, and I removed some dependencies from ACE in `yarp::os`, and one dependency from `yarp::os::impl` in `yarprun`.

I vote for the `yarp::conf` version, but it is slightly less safe and will require some extra testing because it is not just moving one method from one place to another

drdanz commented:
Withdrawing this in favour of #2334
